run_tests.py

```diff
@@ -42,6 +42,28 @@
 else:
     exe_suffix = ''
 
+def setup_pythonpath():
+    # Make sure python can import mesonbuild, even if we change directories as
+    # some tests do. Since sys.path is the final product of fairly complex code
+    # in site.py, it's hard to tell where each entry came from just by looking
+    # at sys.path, so we don't know if a given entry was set from a relative or
+    # absolute path. If an entry was set from a relative path, it won't
+    # continue to work if we change directories.  Instead of trying to guess
+    # where a given entry came from, just add the known-good mesonbuild to
+    # PYTHONPATH so that it will continue to be importable from other
+    # directories.
+    import mesonbuild
```

Comment:
This is still wrong. In the [Debian packaging](https://packages.debian.org/testing/amd64/meson/filelist), `mesonbuild` is not available in the system module path. This means that when you want to run the tests against the system meson, this will import `mesonbuild` from the source tree or fail to find it altogether.

The module to use always depends on the location of `meson`, which is what `detect_meson_py_location()` does.

Of course, it would simplify a lot of things if meson were installed into the system python site-packages on Debian, like everywhere else. @jpakkane?

Reply:
I agree that we should change the Debian packaging to use the system-wide site-packages. The `detect_meson_py_location()` code is quite error-prone and difficult to reason about, and it breaks the wrapper script case that Yocto has (and other tools likely will too). I think it's much more reliable to rely on the Python import system, which already has the right logic to deal with everything.

Reply:
Every time I even glance at Debian packaging I lose 6-10 hours of my life. But maybe we should change it to do the default thing and then just tell people that if they import the internals and anything breaks, they get to keep both pieces.

Reply:
As a maintainer of a few Debian packages myself, I laughed quite hard at that comment :).

Taking a look at the Debian packaging, it looks like the issue comes down to this line in the rules file under the `override_dh_install` target:

`python3 setup.py install --root=$$(pwd)/debian/meson --prefix=/usr --install-layout=deb --install-lib=/usr/share/meson --install-scripts=/usr/share/meson`

If you omit `--install-lib` and `--install-scripts`, I believe it will install into `/usr/lib` instead and everything should work with `sys.path`. Is there a particular reason to put things into `/usr/share/meson` instead?

Reply:
The original reason for that was that Debian packaging insists that every Python package is an importable library with egg info and all that. It also absolutely, positively insists that packages created by it have the name `python3-foobar`. Both of these are bad choices for standalone programs that just happen to be implemented in Python.

The package name should be just `meson`. Pythonness should not be visible in the package name, because it does not provide any "Python thing" for other packages to use. It is just an executable. Similarly we don't want to have Meson importable by any other package, because if it is possible, then someone _will_ do it. Since there is no API stability in the `mesonbuild` package, this means that if the package in question has automatic tests, then new package uploads would get blocked because the tests for that package would fail if we change the internals. This blocks package migrations for Meson at least in Ubuntu, possibly in the future also in Debian.

Reply:
OK, I understand. If this concern is about third-parties importing `mesonbuild` and then relying on a non-committed API, we could rename `mesonbuild` to `_mesonbuild`, which would informally tell people the API is private. Alternatively, the `meson` executable could add `/usr/share/meson` to `PYTHONPATH` as a special-case. Or of course, we could just tell loudly document that the API is not public and ignore people who complain :). It's too bad that, in the end, Python gives you no real mechanisms to stop people.

I think any of these will be more reliable than heuristically trying to find the `mesonbuild` module, which causes other issues (https://github.com/mesonbuild/meson/issues/3015 and this request) and is difficult to maintain.

Reply:
What if, prior to importing, we add `/usr/share/mesonbuild` to `sys.path` in the `meson` and similar executables? Then I think you could satisfy everyone's requirements.

Reply:
What do we need to get this pull request merged?

Reply:
I guess we can change the Debian packaging to do the default thing. Trying to be "smarter" about it is just too difficult. :(

@nirbheek if you think this is good to go, then it's ok to merge.